src/librustc_jobserver/lib.rs

```diff
@@ -0,0 +1,292 @@
+//!
+//! There are 3 primary APIs this crate exposes:
+//!  * acquire()
+//!  * release()
+//!  * acquire_from_request()
+//!  * request_token()
+//!
+//! The first two, acquire and release, are blocking functions which acquire
+//! and release a jobserver token.
+//!
+//! The latter two help manage async requesting of tokens: specifically,
+//! acquire_from_request() will block on acquiring token but will not request it
+//! from the jobserver itself, whereas the last one just requests a token (and
+//! should return pretty quickly, i.e., it does not block on some event).
+//!
+//! -------------------------------------
+//!
+//! We also have two modes to manage here. In the primary (default) mode we
+//! communicate directly with the underlying jobserver (i.e., all
+//! acquire/release requests fall through to the jobserver crate's
+//! acquire/release functions).
+//!
+//! This can be quite poor for scalability, as at least on current Linux
+//! kernels, each release on the jobserver will trigger the kernel to wake up
+//! *all* waiters instead of just one, which, if you have lots of threads
+//! waiting, is quite bad.
+//!
+//! For that reason, we have a second mode which utilizes Cargo to improve
+//! scaling here. In that mode, we have slightly altered communication with the
+//! jobserver. Instead of just blocking on the jobserver, we will instead first
+//! print to stderr a JSON message indicating that we're interested in receiving
+//! a jobserver token, and only then block on actually receiving said token. On
+//! release, we don't write into the jobserver at all, instead merely printing
+//! out that we've released a token.
+//!
+//! Note that the second mode allows Cargo to hook up each rustc with its own
+//! jobserver (i.e., one per rustc process) and then fairly easily make sure to
+//! fulfill the requests from rustc and such. Ultimately, that means that we
+//! have just one rustc thread waiting on the jobserver: a solution that is
+//! nearly optimal for scalability.
+
+use jobserver_crate::Client;
+use lazy_static::lazy_static;
+use rustc_serialize::json::as_json;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::sync::Condvar;
+use std::sync::Mutex;
+
+lazy_static! {
+    // We can only call `from_env` once per process
+
+    // Note that this is unsafe because it may misinterpret file descriptors
+    // on Unix as jobserver file descriptors. We hopefully execute this near
+    // the beginning of the process though to ensure we don't get false
+    // positives, or in other words we try to execute this before we open
+    // any file descriptors ourselves.
+    //
+    // Pick a "reasonable maximum" if we don't otherwise have
+    // a jobserver in our environment, capping out at 32 so we
+    // don't take everything down by hogging the process run queue.
+    // The fixed number is used to have deterministic compilation
+    // across machines.
```

Comment:
I suspect this is copy/pasted from elsewhere, but in any case the last sentence here, the bit about deterministic compilation, isn't actually correct and it's safe to delete that sentence.